blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-        # circumstances is always of type StreamingBody.
-        # Code in botocore.endpoint is responsible for creating the
-        # StreamingBody object. But the result could instead be a string
-        # (if there is an error) or bytes (if handling a different
-        # parameter.)
-        response = self.object.get(Range=range)
-        status = response['ResponseMetadata']['HTTPStatusCode']  # type: ignore
-        if status > 300:
-            raise BlobError(f'Received status code {status} '
-                            f'when getting blob {self.object}')
-        # Make the assumption that the typecheck ignore is making
-        # explicit
-        assert isinstance(response['Body'], StreamingBody), (
-            f"Response from S3 of type {type(response['Body'])}, "
-            f"not StreamingBody.")
-
-        return response['Body'].read()  # type: ignore
-
-    def readinto(self, b: bytearray) -> int:
-        to_read = min(len(self) - self.position, len(b))
-        b[:to_read] = self.read(to_read)
-        return to_read
-
-    def readable(self) -> bool:
-        return True
-
-    def tell(self) -> int:
-        return self.position
-
-    def __repr__(self) -> str:
-        return f'<{type(self).__name__} wrapping {self.object}>'
-
-    def __len__(self) -> int:
-        return self.object.content_length
-
-
-WriterCallback = Callable[[str], None]
-@inherit_docstrings
-class BlobWriter(io.BufferedIOBase):
-    """A file-like object wrapping an S3 blob with an automatic key.
-
-    Add contents to the internal buffer with `write`. No data is
-    transferred until `flush` is called.
-    """
-    def __init__(self, s3_bucket: S3Bucket) -> None:
-        self.bucket = s3_bucket
-        self.checksum = hashlib.sha256()
-        self.buffer = io.BytesIO()
-        self.callbacks: List[WriterCallback] = []
-
-    def register_callback(self, cb: WriterCallback) -> None:
-        """Register the callable cb to be called when the writer is flushed.
-
-        Argument:
-            cb (callable(key)): will be called with whatever key was
-                written to when the writer is flushed.
-        """
-        self.callbacks.append(cb)
-
-    def readable(self) -> bool:
-        return False
-
-    def writable(self) -> bool:
-        if self.closed:
-            raise ValueError(
-                f'I/O operation on closed {type(self).__name__}'
-            )
-
-        return True
-
-    def seekable(self) -> bool:
-        return False
-
-    def write(self, b: bytes) -> int:
-        if self.closed:
-            raise ValueError(
-                f'I/O operation on closed {type(self).__name__}'
-            )
-
-        self.checksum.update(b)
-        self.buffer.write(b)
-
-        return len(b)
-
-    def key(self) -> str:
-        return self.checksum.hexdigest()
-
-    def flush(self) -> None:
-        # FIXME: it doesn't really make sense to call this multiple
```

Comment:
What's the `FIXME` here? This will raise when called after closing.

Reply:
The discrepancy is that if you have a buffered writeable file-like object you can do:
```python3
fd.write('foo')
fd.write('bar')
fd.flush()
fd.write('baz)
fd.flush()
```
and this all makes sense and works and your file then contains `foobarbaz` (and maybe preceding stuff). With S3 this *does not* make sense; you upload entire objects.

We could instead make the `Writer` act more like a buffer; `flush()` is a NOP and `close()` writes everything. This matches the file-like API better in one way (multiple `flush()`es do nothing rather than raise exceptions) but worse in that semantically, `flush()` is what causes a write, not `close()` (but `close` calls `flush`.)

Note that with the current implementation it is not at all possible to upload anything to S3 before all the data has been written to the `Writer`, because it must all have been seen to generate the key.